Adds new SecHUD designations and tweaks secHUD on examining (updated 22 May)

Regular-Joe-SF opened the pull request:
This is a new take on https://github.com/ParadiseSS13/Paradise/pull/11909, with an additional feature.

**What does this PR do:**
This PR adds new criminal_status designations, along with respective secHUD icons:
- Search
- Demote
- Monitor.

And alters info provided by secHUD on examination, so, that it will now show the following:
- Criminal status (no changes) 
- Security records:  on click, spits out the comment log, and preserves comment adding on the fly
- Latest entry: gives out the latest entry to the comment log

Backend change:
- Criminal statuses are now defines, instead of the values given separately each time using them.

**Aim of the change**
Twofold idea behind this:
-  in general, enchance the basic tool of security, in order to make security better in their jobs.

1. On new criminal status designations, to provide a visual aid for officers on determining which kind of treatment each each crewmember, marked by security, is to receive. 

> Presently, there are two criminal status designations, execute and arrest, and while the usage of execution is obvious, crewmembers marked with arrest status are to be treated quite differently. 

>Search designation would mark a search warrant, which is determined at sec SOP. Personnel marked with search warrant are to be searched and released, if no reason to arrest them would take place.

>Demote warrant would mark sole escort to demote -mission, which are quite redundant. This was suggested by Kyet on my first take on the matter.

>Monitor warrant would help to visualize concerns and suspicions, that are no grounds for arrests or relevant to tag with search. I thought undercover antags, like changelings by usual, with this, and the situations in which sec has no good leads, but lots of concerns. Presently I've seen people to use the released tag for these purposes. Monitor-tagging would help officers to keep eye on such at the field. This would be particularly so during high-pop and-or otherwise busy rounds, in which it is hard to keep up and remember cases, that would need more investigation, rather than direct action.

2. On altering the secHUD, to make the basic tool more convenient than it is presently; while examing a person with it, the comment log is behind two buttons, making it slow to access in rapid situations and also a commonly unknown feature. 

>The new "latest entry" -feature will provide the latest comment on the "comment" log, on examination. 

>If target has no crew/sec records, or we can't recognize the person, it says: "ERROR: Unable to locate a data core entry for this person." 

>If target is crew but no comments are present, it says: "No entries."

>If target has comment log modified, it says the most recent comment.

>Kyet brought this up in the previous one, so I went on thinking this. Security records in common would need refactoring: presently there are "security records", which consists of multiple single-status fields, that won't log changes to them - I doubt if this is of any good use. Seems like ancient thing. Then, there are "comments", which are queried each time criminal_status is modified, and when a comment is added or deleted otherwise. Practically sec recors now are the "comments."

>That for, this makes the HUD to provide buttons to view the "comment" log in complete, and to add new "comments" on examination - removing the access to the "sec records", which are obsolete.

**Minor things - done along changing the usage of criminal status**
- Turrets, which have their own setup known as "assess_perp", didn't have the line to arrest executees. I set the value for execute status to 7 similar to beepsky "judgebot". Remote example would be a RD with execute status accessing AI upload in all peace. 

- Beepsky will now arrest people with Incarcerated status 

- CC 'evilfax' answer option 'Demote' will utilize the Demote status, and both Demote and Demote with bot will give the demotion order to the sec comments.

- Search, Demote and Monitor won't provide additional threatlevel.

- Execution HUD tag now has mildly flashing animation. To visually distinguish it from the other red tags.

- Ghosts can now see the info sechud provides via examining, if they have all huds or sechud enabled, including using the view comment log button.


**To be considered**
- Thanks for help on the way - this was a noob project, I have learned things while doing. Will appreciate any comments on conducting this, what could have done better/be done better now! 



**Images of sprite changes:**
![K00_New-warrants](https://user-images.githubusercontent.com/53127823/67817830-affc0880-fab7-11e9-849a-5d6beff7accf.PNG)
![K0_if-unknown](https://user-images.githubusercontent.com/53127823/67817840-ba1e0700-fab7-11e9-8ae8-4216cd700987.PNG)
![K1_No-entry](https://user-images.githubusercontent.com/53127823/67817843-bd18f780-fab7-11e9-949c-28e91ad54588.PNG)
![K2_Latest entry](https://user-images.githubusercontent.com/53127823/67817844-c013e800-fab7-11e9-858f-b97b05e47170.PNG)
![K3-comment_log](https://user-images.githubusercontent.com/53127823/67817847-c2764200-fab7-11e9-8dcb-420476527225.PNG)

🆑 Regular-Joe-SF
add: *New Criminal Status designations: Search, Demote, Monitor
add: *New secHUD examination function: Latest security comment visible instantly
tweak: Examining with secHUD now provides instantly the "view the comment log" ability, instead of obsolete "view security records"
tweak: Security record criminal statuses are now defines
tweak: CC evilfax option 'Demote' utilizes Demote status and adds the demotion order to sec comments
remove: Examining with the secHUD will no longer give the button to view the "security records." 
fix: Fixed turrets not regarding the execute status
imageadd: Hud sprites for Search, Demote and Monitor
imagetweak: Mildly flashing animation for the execution hudsprite
/🆑

farie82 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -340,16 +340,32 @@
 					for(var/datum/data/record/R in data_core.security)
 						if(R.fields["id"] == E.fields["id"])
 							criminal = R.fields["criminal"]
+			if(commentLatest)
```
This does nothing since itll always be defined as "None" here

xXx-RegularJoe-xXx replied:
I figured to define it that way, as of it does provide the "None" message if there is no datacore entry for the examined person, that is, it is unknown or not-crew. I guess it works so, that if the list macros return 0, as of the list in question is not existing, it will state the defined string. But if there is the list, it will return either latest comment or no entries in case of list existing, but it is empty. Comment log is defined at datacore.dm.

Could change the "None" to be more informative though, like "ERROR: Unable to locate a data core entry for this person", which is the message when you try to view comment log of a unknown person.

farie82 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -340,16 +340,32 @@
+					if(E.fields["name"] == perpname)
+						for(var/datum/data/record/R in data_core.security)
+							if(R.fields["id"] == E.fields["id"])
+								if(LAZYLEN(R.fields["comments"]))
```
Put this in the loop done just above. It also searches the perp in the db and then assigns the criminal status. Best to search the last comment status in the same loop

xXx-RegularJoe-xXx replied:
Done!

farie82 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -340,16 +340,32 @@
+									commentLatest = "No entries."
+
 			var/criminal_status = hasHUD(user, "read_only_security") ? "\[[criminal]\]" : "<a href='?src=[UID()];criminal=1'>\[[criminal]\]</a>"
+			var/commentLatest_check = "[commentLatest]"
```
This is a copy of the value. If you want [comment] as text you should use \[comment\]

You can also put this directly in the msg like the other info things do

xXx-RegularJoe-xXx replied:
True that! The difference between machine learning and monkey learning, the monkey learns just slower...

Thanks for these!

farie82 commented:
Good stuff this man. Always annoying to have to do so many clicks to get what you want to see

pinatacolada commented:
monitor icon looks like a chinese bootleg mcdonalds logo and i cant unsee it now

TDSSS commented:
Glad to see this brought back, good idea :+1:

xXx-RegularJoe-xXx commented:
> 
> 
> monitor icon looks like a chinese bootleg mcdonalds logo and i cant unsee it now
What if I told you, it is the wanted icon upside down... chinese bootleg fast food criminals everywhere.

xXx-RegularJoe-xXx commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -340,16 +340,26 @@
 						if(R.fields["id"] == E.fields["id"])
 							criminal = R.fields["criminal"]
+							if(LAZYLEN(R.fields["comments"]))
+								for(var/c in R.fields["comments"])
```
There!

farie82 commented on code/game/machinery/machinery.dm:
```diff
@@ -547,6 +547,9 @@ Class Procs:
 		if(check_arrest && R && (R.fields["criminal"] == "*Arrest*"))
 			threatcount += 4
 
+		if(check_arrest && R && (R.fields["criminal"] == "*Execute*" ))
```
You can combine this line with 547 so you dint check check_arrest and R twice. Can even use a switch for that to improve performance but thats not required

xXx-RegularJoe-xXx replied:
There they are now! Noticed by the way that these aren't actually beepsky's lines, but portable turret's behaviour is altered by these at portable_turret.dm. Probably one could just fix the turrets use beepsky's "judgebot" instead, but there they are for now

farie82 commented on code/game/machinery/machinery.dm:
```diff
@@ -539,13 +539,13 @@ Class Procs:
-
 		var/datum/data/record/R = find_security_record("name", perpname)
-		if(check_records && !R)
-			threatcount += 4
```
Nearly perfect. We're missing this threat. When beepsky checks records and the person has no record

xXx-RegularJoe-xXx replied:
Dang, now I just missed it.

farie82 commented on code/game/machinery/machinery.dm:
```diff
@@ -539,13 +539,13 @@ Class Procs:
+				if("*Execute*")
+					threatcount += 7
+				if("*Arrest*")
+					threatcount += 5
```
Threatcount is 5 instead of 4. Prob unintended

xXx-RegularJoe-xXx replied:
Actually intended, as in the beepsky code at /living/human.dm arrest is for 5, so that tweaking this file won't change beepsky behaviour, but turrets.

SabreML commented:
I think the Demote sprite should be a different colour, I can see it being confused with Arrest or Execute.

Vivalas commented:
I love this and sad I'm late since armed and dangerous seemed pretty useful. It doesn't have to have any difference in treatment, just let sec know on sight the difference between greytide and more dangerous people or eocs. But that's all I'll say on that, this is still pretty wicked.

I like the idea of people stealing sechuds now and getting anxious when a m suddenly pops up for them. Big NT is watching.

Also, just a thought: perhaps search could be a yellow magnifying glass, and I dunno what for demote, but symbols might look cool.

xXx-RegularJoe-xXx commented:
Time has passed, here again!

> 
> 
> I think the Demote sprite should be a different colour, I can see it being confused with Arrest or Execute.

When I tried the colors I thought the same, but couldn't get anything better than red off my head. Blue reserved for released and monitor, yellow to search, brown to prisoners, and using lighter variants of the colors present in huds doesn't work well, so went out of ideas myself.


> 
> 
> I love this and sad I'm late since armed and dangerous seemed pretty useful. It doesn't have to have any difference in treatment, just let sec know on sight the difference between greytide and more dangerous people or eocs. But that's all I'll say on that, this is still pretty wicked.
> 
> I like the idea of people stealing sechuds now and getting anxious when a m suddenly pops up for them. Big NT is watching.
> 
> Also, just a thought: perhaps search could be a yellow magnifying glass, and I dunno what for demote, but symbols might look cool.
> 
> EDIT: and oh, my. that latest entry thing is a god send, thank you so much. that will help newer officers out so much

Thanks! And actually, neat idea the hourglass - how didn't I come to it? Indeed! I scribbled this, would it do? Also made the background yellow go transparent in two stages for a minor flashing animation, just because felt like it would be nice - but would a solid one be better?

![searchicon-sechud](https://user-images.githubusercontent.com/53127823/74850021-1dbb9100-5342-11ea-8dcf-26ce1988c722.gif)

taukausanake commented:
As of posting it looks like IconDiffBot has an error. Just like to mention that

AffectedArc07 commented:
IDB fixed. Re-commit when you want to re-generate diffs

Kyep commented:
Could you please make it so that if you're a ghost, and you examine someone who is wanted, it states why they are wanted?
Ghosts can't view records, it'd be helpful.

xXx-RegularJoe-xXx commented:
> 
> 
> It would be great if, when examining someone who was wanted as a ghost, their examine text showed why they were wanted.

Done! Ghosts can now see the info sechud provides via examining, if they have all huds or sechud enabled, including using the view comment log button.

xXx-RegularJoe-xXx commented:
Changed files are now at LF format and conflicts are solved, as of the situation of 11 April!

Kyep commented:
@xXx-RegularJoe-xXx could you solve the merge conflicts, and update from master? I'd like to testmerge this but I can't since its out of date and I don't know who you are on Discord.

xXx-RegularJoe-xXx commented:
Wilco! Expect in next 12hrs. Regular_Joe at discord

farie82 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -343,16 +343,24 @@
 							criminal = R.fields["criminal"]
+							if(LAZYLEN(R.fields["comments"])) //if the commentlist is present
+								var/list/comments = R.fields["comments"]
+								commentLatest = LAZYACCESS(comments,comments.len) //get the latest entry from the comment log
```
spacing

Regular-Joe-SF replied:
Check!

farie82 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -404,6 +412,14 @@
+		if(O.data_hud_seen == DATA_HUD_SECURITY_ADVANCED || O.data_hud_seen == DATA_HUD_DIAGNOSTIC + DATA_HUD_SECURITY_ADVANCED + DATA_HUD_MEDICAL_ADVANCED)
+			switch(hudtype)
+				if("security")
+					return 1
```
TRUE and FALSE below. Best to use those defines if you're using a boolean value

Regular-Joe-SF replied:
Check, thanks!

Regular-Joe-SF commented:
Per Farie's recommend, I moved the criminal statuses to defines, and updated the usage of them accordingly. The new defines are at _DEFINES/criminal_status.dm Minor changes done in the same process:

- CC 'evilfax' option 'demote' will now utilize the Demote status, as well as spitting the demotion order to sec comments
- Beepsky arrests people with Incarcerated status (threat level 5)

farie82 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1554,13 +1554,13 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
-					threatcount += 2
-				if("Parolled")
+				if(SEC_RECORD_STATUS_INCARCERATED)
+					threatcount += 5
```
Might want to check this to see if beepsky doesn't go ham on prisoners who are in cells already

Regular-Joe-SF replied:
Indeed, now I remember realising that previously, already forgot it. Changed it back to 2 as it was. Thanks!

ZomgPonies commented:
@Regular-Joe-SF can you merge master please? I'd like to see this happen sooner rather than later.

Regular-Joe-SF commented:
Done - in case this was a joke for the simple man, well, done, too :D

Crazylemon64 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -768,7 +768,7 @@
 	if(href_list["secrecordComment"])
 		if(hasHUD(usr,"security"))
-			if(usr.incapacitated())
+			if(usr.incapacitated() == !isobserver(usr))
```
What is this supposed to be?
Presumably it's supposed to account for ghosts being counted as "incapacitated"?
However, I don't think this is intuitive enough to go without a comment explaining the what/why here

Regular-Joe-SF replied:
I thought that is a check whether one is incapacitated OR a ghost, and if so, render unable to use secrecordComment even if they can see it. Without this ghosts could add comments by clicking the add comment button from the examine info given by sechud.

Crazylemon64 replied:
Isn't this one in particular a purely informational topic call? Or do you want to lock ghosts out anyways

Regular-Joe-SF replied:
Oh damn, bear with me, I checked it too quickly yesterday. So actually: the check in question _gives_ the access for ghosts to view the comment log by being able to press that button. While adding comments from there is handled by the next href_list "secrecordadd", into which ghosts do not have access as they are 'incapacitated' as well as same renders them unable to interact with the other buttons present.

Re-adding it with a comment - I guess this is exactly why one should do them comments...

Crazylemon64 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -395,17 +403,25 @@
+				return FALSE
+	else if(istype(M, /mob/dead/observer))
+		var/mob/dead/observer/O = M
+		if(O.data_hud_seen == DATA_HUD_SECURITY_ADVANCED || O.data_hud_seen == DATA_HUD_DIAGNOSTIC + DATA_HUD_SECURITY_ADVANCED + DATA_HUD_MEDICAL_ADVANCED)
```
Wouldn't it make more sense to do `O.data_hud_seen & DATA_HUD_SECURITY_ADVANCED` here?

Crazylemon64 replied:
wait no, it looks like we have the ghost "omni-hud" work as a special identifier derived from adding them together.
why do we do it this way this is so cursed
I guess this will have to do for now

Crazylemon64 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -395,17 +403,25 @@
 			else
-				return 0
+				return FALSE
+	else if(istype(M, /mob/dead/observer))
```
`isobserver(M)`

Crazylemon64 commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -768,7 +768,7 @@
 	if(href_list["secrecordComment"])
 		if(hasHUD(usr,"security"))
-			if(usr.incapacitated())
+			if(usr.incapacitated() == !isobserver(usr)) //give the ghosts access to "View Comment Log" while they can't manipulate it
```
Actually, hold on, I'm overlooking a more obvious solution here: `usr.incapacitated() && !isobserver(usr)`, since ghosts don't ever change their incapacitated state (and if they ever lost incapacitation somehow I don't see why they should stop being able to see this)

Regular-Joe-SF replied:
True that, better so. Changed
